lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+    non-tail-position, and it gets rewritten into tail-calls. TMC
+    handles all such cases where the continuation of the call
+    (what needs to be done after the return) is a "construction", the
+    creation of a (possibly nested) data block.
+
+    The code transformation generates two versions of the
+    input function, the "direct" version with the same type and
+    behavior as the original one (here just [map]), and
+    the "destination-passing-style" version (here [map_dps]).
+
+    Any call to the original function from outside the let..rec
+    declaration gets transformed into a call into the direct version,
+    which will itself call the destination-passing-style versions on
+    recursive calls that may benefit from it (they are in tail-position
+    modulo constructors).
+
+    Because of this inherent code duplication, the transformation may
+    not always improve performance. In this implementation, TMC is
+    opt-in, we only transform functions that the user has annotated
+    with an attribute to request the transformation.
+*)
+
+type 'offset destination = {
+  var: Ident.t;
+  offset: 'offset;
+  loc : Debuginfo.Scoped_location.t;
+}
+and offset = lambda
+(** In the OCaml value model, interior pointers are not allowed.  To
+    represent the "placeholder to mutate" in DPS code, we thus use a pair
+    of the block containing the placeholder, and the offset of the
+    placeholder within the block.
+
+    In the common case, this offset is an arbitrary lambda expression, typically
+    a constant integer or a variable. We define ['a destination] as parametrized
+    over the offset type to represent formal destination parameters (where
+    the offset is an Ident.t), and maybe in the future statically-known
+    offsets (where the offset is an integer).
+*)
+
+let add_dst_params ({var; offset} : Ident.t destination) params =
+  (var, Pgenval) :: (offset, Pintval) :: params
+
+let add_dst_args ({var; offset} : offset destination) args =
+  Lvar var :: offset :: args
+
+let assign_to_dst {var; offset; loc} lam =
+  Lprim(Psetfield_computed(Pointer, Heap_initialization),
+        [Lvar var; offset; lam], loc)
+
+(** The type [Constr.t] represents a reified constructor with a single hole, which can
+    be either directly applied to a [lambda] term, or be used to create
+    a fresh [lambda destination] with a placeholder.
+ *)
+module Constr = struct
+  type t = {
+    tag : int;
+    flag: Asttypes.mutable_flag;
+    shape : block_shape;
+    before: lambda list;
+    after: lambda list;
+    loc : Debuginfo.Scoped_location.t;
+  }
+
+  let apply ?flag con t =
+    let flag = match flag with None -> con.flag | Some flag -> flag in
+    let block_args = List.append con.before @@ t :: con.after in
+    Lprim (Pmakeblock (con.tag, flag, con.shape), block_args, con.loc)
+
+  let tmc_placeholder = Lconst (Const_base (Const_int 0))
+  (* TODO consider using a more magical constant like 42, for debugging? *)
```

Comment:
https://en.wikipedia.org/wiki/Magic_number_(programming)#Debug_values - this is a good list of Debug values. ;-)